src/pycldf/dataset.py

```diff
@@ -302,20 +303,23 @@ def from_metadata(cls, fname):
     @classmethod
     def from_data(cls, fname):
         fname = Path(fname)
-        for mod in get_modules():
-            if mod.match(fname):
-                res = mod.cls.from_metadata(fname if fname.is_dir() else fname.parent)
-                for line in fname.open(encoding='utf8'):
-                    required_cols = [
-                        c.name for c in res[res.primary_table].tableSchema.columns
-                        if c.required]
-                    if not set(required_cols).issubset(set(line.split(','))):
-                        raise ValueError()
-                    break
-                else:
-                    raise ValueError('empty data file!')
-                return res
-        raise ValueError(fname)
+        colnames = next(iterrows(fname), [])
+        if not colnames:
+            raise ValueError('empty data file!')
+        if cls is Dataset:
+            try:
+                cls = next(mod.cls for mod in get_modules() if mod.match(fname))
+            except StopIteration:
+                raise ValueError(fname)
+            assert issubclass(cls, Dataset) and cls is not Dataset
+
+        res = cls.from_metadata(fname if fname.is_dir() else fname.parent)
```

Comment:
@xrotwang: this is always `False`, right (reading the column names implies that `fname` is a file)?

Reply:
Yes. I think at one point I had the idea of looking for any recognizable file within a directory. But now I'd say that's way too  much guessing.

Reply:
Thanks, will simplify accordingly.